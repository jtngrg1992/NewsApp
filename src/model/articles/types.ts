export type ArticleMedia = {
  thumbnail?: string;
  mediumImage?: string;
  largeImage?: string;
};

export type Article = {
  id: number;
  url: string;
  source: string;
  published: string;
  updated: string;
  author: string;
  title: string;
  abstract: string;
  media?: ArticleMedia;
};

export type ArticlesState = {
  list: Article[];
  isFetchingArticles: boolean;
  error?: Error;
};

export const FETCH_ARTICLES = 'FETCH_ARTICLES';
export const FETCH_ARTICLES_SUCCESS = 'FETCH_ARTICLES_SUCCESS';
export const FETCH_ARTICLES_FAILURE = 'FETCH_ARTICLES_FAILURE';
export const FETCH_ARTICLES_PENDING = 'FETCH_ARTICLES_PENDING';

export type FetchArticlesAction = {
  type: typeof FETCH_ARTICLES;
};

export type FetchArticlesPendingAction = {
  type: typeof FETCH_ARTICLES_PENDING;
};

export type FetchArticlesSuccessAction = {
  type: typeof FETCH_ARTICLES_SUCCESS;
  payload: Article[];
};

export type FetchArticlesFailureAction = {
  type: typeof FETCH_ARTICLES_FAILURE;
  error: Error;
};

export type ArticlesActionTypes =
  | FetchArticlesAction
  | FetchArticlesSuccessAction
  | FetchArticlesFailureAction
  | FetchArticlesPendingAction;
