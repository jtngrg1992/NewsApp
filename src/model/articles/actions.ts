import {
  FetchArticlesAction,
  FETCH_ARTICLES,
  FETCH_ARTICLES_PENDING,
  FETCH_ARTICLES_SUCCESS,
  FETCH_ARTICLES_FAILURE,
  FetchArticlesPendingAction,
  FetchArticlesFailureAction,
  FetchArticlesSuccessAction,
  Article,
} from './types';

export const fetchArticles = (): FetchArticlesAction => ({
  type: FETCH_ARTICLES,
});

export const fetchArticlesPending = (): FetchArticlesPendingAction => ({
  type: FETCH_ARTICLES_PENDING,
});

export const fetchArticlesSuccess = (
  articles: Article[],
): FetchArticlesSuccessAction => ({
  type: FETCH_ARTICLES_SUCCESS,
  payload: articles,
});

export const fetchArticlesFailure = (
  error: Error,
): FetchArticlesFailureAction => ({
  type: FETCH_ARTICLES_FAILURE,
  error,
});
