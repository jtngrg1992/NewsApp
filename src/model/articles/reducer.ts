import {
  ArticlesState,
  ArticlesActionTypes,
  FETCH_ARTICLES_PENDING,
  FETCH_ARTICLES_SUCCESS,
  FETCH_ARTICLES_FAILURE,
} from './types';

const initialState: ArticlesState = {
  list: [],
  isFetchingArticles: false,
};

export default (
  state = initialState,
  action: ArticlesActionTypes,
): ArticlesState => {
  switch (action.type) {
    case FETCH_ARTICLES_PENDING:
      return {
        ...state,
        isFetchingArticles: true,
        error: undefined,
      };
    case FETCH_ARTICLES_SUCCESS:
      return {
        ...state,
        list: action.payload,
        isFetchingArticles: false,
      };
    case FETCH_ARTICLES_FAILURE:
      return {
        ...state,
        error: action.error,
        isFetchingArticles: false,
      };
    default:
      return state;
  }
};
