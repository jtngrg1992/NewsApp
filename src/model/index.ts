import {combineReducers} from 'redux';
import {all} from 'redux-saga/effects';
import articlesReducer from './articles/reducer';
import watchFetchArticles from './articles/sagas';

export const rootReducer = combineReducers({articles: articlesReducer});

export function* rootSaga() {
  yield all([watchFetchArticles()]);
}
