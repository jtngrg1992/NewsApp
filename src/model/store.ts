import {rootReducer, rootSaga} from './index';
import {applyMiddleware, createStore} from 'redux';
import createSagaMiddleware from 'redux-saga';

const sagaMiddleware = createSagaMiddleware();
const middlewares = applyMiddleware(sagaMiddleware);
const store = createStore(rootReducer, middlewares);

sagaMiddleware.run(rootSaga);

export {store};

export type RootState = ReturnType<typeof rootReducer>;
