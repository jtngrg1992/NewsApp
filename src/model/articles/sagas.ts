import {takeEvery, call, put} from 'redux-saga/effects';
import {fetchArticlesSuccess, fetchArticlesFailure} from './actions';
import {FETCH_ARTICLES, Article, ArticleMedia} from './types';

export const processArticleJSON = (articleJSON: any[]): Article[] => {
  return articleJSON.map((item) => {
    const {
      id,
      media,
      published_date,
      updated,
      title,
      url,
      abstract,
      source,
      byline,
    } = item;

    let articleMedia: ArticleMedia | undefined;

    if (media && Array.isArray(media)) {
      const m = media as any[];

      if (m.length > 0) {
        const mediaMeta = m[0]['media-metadata'] as any[];
        if (mediaMeta) {
          const thumbnail = mediaMeta.find(
            (i) => i.format === 'Standard Thumbnail',
          )?.url;
          const mediumImage = mediaMeta.find(
            (i) => i.format === 'mediumThreeByTwo210',
          )?.url;
          const largeImage = mediaMeta.find(
            (i) => i.format === 'mediumThreeByTwo440',
          )?.url;

          articleMedia = {
            thumbnail,
            mediumImage,
            largeImage,
          };
        }
      }
    }

    let itemToReturn: Article = {
      id,
      url,
      published: published_date,
      title,
      abstract,
      source,
      author: byline,
      updated,
      media: articleMedia,
    };

    return itemToReturn;
  });
};

export const ARTICLE_URL =
  'https://api.nytimes.com/svc/mostpopular/v2/viewed/7.json?api-key=WvQCGKUDhY2YSxG5K4hS2bgKC0eD52Og';
export function* getArticles() {
  try {
    const response = yield call(fetch, ARTICLE_URL);
    const responseJSON = yield response.json();
    const {results} = responseJSON;
    const processedResults = processArticleJSON(results);
    yield put(fetchArticlesSuccess(processedResults));
  } catch (e) {
    yield put(fetchArticlesFailure(e));
  }
}

function* watchFetchArticles() {
  yield takeEvery(FETCH_ARTICLES, getArticles);
}

export default watchFetchArticles;
