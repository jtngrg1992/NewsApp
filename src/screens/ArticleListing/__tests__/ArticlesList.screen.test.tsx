import React from 'react';
import {shallow, ShallowWrapper} from 'enzyme';
import ArticlesListScreen from '@news/screens/ArticleListing/ArticlesList.screen';
import {Article, fetchArticles} from '@news/model/articles';
import mockArticles from '../../../model/articles/mock-data';
import {FlatListProps, ListRenderItem} from 'react-native';
import {useNavigation} from '@react-navigation/native';

jest.mock('@news/hooks', () => ({
  useArticleList: () => mockArticles,
  useIsFetchingArticles: () => true,
  useArticlesFetchError: () => new Error('Mock Error'),
}));

const mockSpy = jest.fn();
jest.mock('react-redux', () => ({
  // @ts-ignore
  ...jest.requireActual('react-redux'),
  useDispatch: () => mockSpy,
}));

describe('unit/snapshot tests for ArticlesListScreen', () => {
  let wrapper: ShallowWrapper;

  beforeEach(() => {
    wrapper = shallow(<ArticlesListScreen />);
  });

  afterEach(() => {
    mockSpy.mockClear();
  });

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should ask for articles after coming in focus', () => {
    expect(mockSpy).toHaveBeenCalledWith(fetchArticles());
  });

  it('should render a flatlist for displaying article rows', () => {
    expect(wrapper.find('FlatList')).toBeDefined();
  });

  it('should render all the articles that are present in the state', () => {
    expect((wrapper.props() as FlatListProps<any>).data?.length).toBe(
      mockArticles.length,
    );
  });

  it('should appear refreshing if the articles are being fetched', () => {
    expect((wrapper.props() as FlatListProps<any>).refreshing).toBeTruthy();
  });

  it('should navigate to detail screen if an article is tapped upon', () => {
    const props = wrapper.props() as FlatListProps<any>;
    const renderItem: ListRenderItem<Article> | null | undefined =
      props.renderItem;
    const row =
      renderItem &&
      renderItem({
        item: mockArticles[0],
        index: 0,
        separators: {
          highlight: () => {},
          unhighlight: () => {},
          updateProps: () => {},
        },
      });
    const spy = jest.spyOn(useNavigation(), 'navigate');
    row?.props.onPress();
    expect(spy).toHaveBeenCalledWith('Detail', {article: mockArticles[0]});
  });
});
