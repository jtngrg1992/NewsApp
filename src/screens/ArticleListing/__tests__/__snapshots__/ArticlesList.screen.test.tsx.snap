// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`unit/snapshot tests for ArticlesListScreen should render correctly 1`] = `
<FlatList
  data={
    Array [
      Object {
        "abstract": "Mock Abstract",
        "author": "Mock Author",
        "id": 0,
        "media": Object {
          "largeImage": "some.other.url",
          "mediumImage": undefined,
          "thumbnail": "some.url",
        },
        "published": "2021-01-05",
        "source": "Mock source",
        "title": "Mock title",
        "updated": "2021-01-05",
        "url": "mock url",
      },
      Object {
        "abstract": "Mock Abstract2",
        "author": "Mock Author",
        "id": 1,
        "media": Object {
          "thumbnail": "some.url",
        },
        "published": "2021-01-05",
        "source": "Mock source",
        "title": "Mock title2",
        "updated": "2021-01-05",
        "url": "mock url",
      },
    ]
  }
  disableVirtualization={false}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  numColumns={1}
  onEndReachedThreshold={2}
  refreshControl={
    <RefreshControlMock
      refreshing={true}
    />
  }
  refreshing={true}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  style={
    Object {
      "flex": 1,
    }
  }
  updateCellsBatchingPeriod={50}
  windowSize={21}
/>
`;
